modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java

```diff
@@ -298,10 +356,106 @@ void countRetry() {
         retries.incrementAndGet();
     }
 
-    public void countThrottle(TimeValue delay) {
-        long nanos = delay.nanos();
-        if (nanos > 0) {
-            throttledNanos.addAndGet(nanos);
+    float getRequestsPerSecond() {
+        return requestsPerSecond;
+    }
+
+    /**
+     * Schedule prepareBulkRequest to run after some delay. This is where throttling plugs into reindexing so the request can be rescheduled
+     * over and over again.
+     */
+    void delayPrepareBulkRequest(ThreadPool threadPool, TimeValue delay, AbstractRunnable prepareBulkRequest) {
+        // Synchronize so we are less likely to schedule the same request twice.
+        synchronized (delayedPrepareBulkRequest) {
+            delayedPrepareBulkRequest.set(new DelayedPrepareBulkRequest(threadPool, getRequestsPerSecond(), delay, new AbstractRunnable() {
+                private final AtomicBoolean oneTime = new AtomicBoolean();
+
+                @Override
+                protected void doRun() throws Exception {
+                    // Paranoia to prevent furiously rethrottling from running the command multiple times. Without this we totally can.
+                    if (false == oneTime.getAndSet(true)) {
```

Comment:
This might read better if it were renamed to `hasRun` initialized with false and then it could be

```java
if (hasRun.compareAndSet(false, true)) {
   prepareBulkRequest.run();
}
```